envoy/network/transport_socket.h

```diff
@@ -252,10 +257,12 @@ class TransportSocketFactory {
 
   /**
    * @param options for creating the transport socket
+   * @param host description for the destination upstream host
    * @return Network::TransportSocketPtr a transport socket to be passed to connection.
    */
   virtual TransportSocketPtr
-  createTransportSocket(TransportSocketOptionsConstSharedPtr options) const PURE;
+  createTransportSocket(TransportSocketOptionsConstSharedPtr options,
+                        std::shared_ptr<const Upstream::HostDescription> host) const PURE;
```

Comment:
It is undesired that transport socket will have dependency on upstream while this interface is not specific to upstream. Since we only need filter state and metadata here, why not use the filterState in TransportSocketOptions?

Reply:
The filter state is already in the options, so supporting just the filter state does not need any interface changes.

How would I propagate the endpoint metadata to the transport socket? I don't think we can write a downstream filter easily that copies metadata into the filter state, since the upstream owns the host selection.

Reply:
Yeah that actually makes more sense.

Reply:
SG. I'd like to split the refactor to minimize code changes. It seems to touch tons of files. Will push a separate PR to split downstream and upstream transport sockets.

Reply:
Filed https://github.com/envoyproxy/envoy/pull/21648, slogging through test build fixes.